src/mesh/coordinates.cxx

```diff
@@ -17,6 +17,127 @@
 
 #include <globals.hxx>
 
+// use anonymous namespace so this utility function is not available outside this file
+namespace {
+/// Interpolate a Field2D to a new CELL_LOC with interp_to.
+/// Communicates to set internal guard cells.
+/// Boundary guard cells are set by extrapolating from the grid, like
+/// 'free_o3' boundary conditions
+/// Corner guard cells are set to BoutNaN
+Field2D interpolateAndExtrapolate(const Field2D& f, CELL_LOC location,
+    bool extrapolate_x = true, bool extrapolate_y = true) {
+
+  Mesh* localmesh = f.getMesh();
+  Field2D result = interp_to(f, location, RGN_NOBNDRY);
+  // Ensure result's data is unique. Otherwise result might be a duplicate of
+  // f (if no interpolation is needed, e.g. if interpolation is in the
+  // z-direction); then f would be communicated. Since this function is used
+  // on geometrical quantities that might not be periodic in y even on closed
+  // field lines (due to dependence on integrated shear), we don't want to
+  // communicate f. We will sort out result's boundary guard cells below, but
+  // not f's so we don't want to change f.
+  result.allocate();
+  localmesh->communicate(result);
+
+  // Extrapolate into boundaries (if requested) so that differential geometry
+  // terms can be interpolated if necessary
+  // Note: cannot use applyBoundary("free_o3") here because applyBoundary()
+  // would try to create a new Coordinates object since we have not finished
+  // initializing yet, leading to an infinite recursion.
+  // Also, here we interpolate for the boundary points at xstart/ystart and
+  // (xend+1)/(yend+1) instead of extrapolating.
+  for (auto bndry : localmesh->getBoundaries()) {
+    if ((extrapolate_x and bndry->bx != 0) or (extrapolate_y and bndry->by != 0)) {
+      int extrap_start = 0;
+      if ((location == CELL_XLOW) && (bndry->bx > 0)) {
+        extrap_start = 1;
+      } else if ((location == CELL_YLOW) && (bndry->by > 0)) {
+        extrap_start = 1;
+      }
+      for (bndry->first(); !bndry->isDone(); bndry->next1d()) {
+        // interpolate extra boundary point that is missed by interp_to, if
+        // necessary.
+        // Only interpolate this point if we are actually changing location. E.g.
+        // when we use this function to extrapolate J and Bxy on staggered grids,
+        // this point should already be set correctly because the metric
+        // components have been interpolated to here.
+        if (extrap_start > 0 and f.getLocation() != location) {
+          // note that either bx or by is >0 here
+          result(bndry->x, bndry->y) =
+              (9. * (f(bndry->x - bndry->bx, bndry->y - bndry->by) + f(bndry->x, bndry->y))
+               - f(bndry->x - 2 * bndry->bx, bndry->y - 2 * bndry->by)
+               - f(bndry->x + bndry->bx, bndry->y + bndry->by))
```

Comment:
It does require at least 2 guard cells, the same as `interp_to` so this should always be true for a direction in which some field is staggered, but I'll add ASSERTs to check. In the direction into the grid, it's OK because it only needs 2 points inside the boundary; there must be at least one grid point and one guard cell (since if there are no guard cells there is no boundary cell to fill), so these two points are always available (if this expression is evaluated at all).